stl/inc/atomic

```diff
@@ -131,6 +131,68 @@ _NODISCARD extern "C" bool __cdecl __std_atomic_has_cmpxchg16b() noexcept;
 #define ATOMIC_LLONG_LOCK_FREE    2
 #define ATOMIC_POINTER_LOCK_FREE  2
 
+#if _HAS_CXX20
+extern "C" {
+void __stdcall __std_atomic_wait_direct(const void* const _Storage, const void* const _Comparand,
+    const std::size_t _Size, unsigned long long& _Wait_context) noexcept;
+void __stdcall __std_atomic_notify_one_direct(const void* const _Storage) noexcept;
+void __stdcall __std_atomic_notify_all_direct(const void* const _Storage) noexcept;
+
+void __stdcall __std_atomic_unwait_direct(const void* const _Storage, unsigned long long& _Wait_context) noexcept;
+
+void __stdcall __std_atomic_wait_indirect(const void* const _Storage, unsigned long long& _Wait_context) noexcept;
+void __stdcall __std_atomic_notify_one_indirect(const void* const _Storage) noexcept;
+void __stdcall __std_atomic_notify_all_indirect(const void* const _Storage) noexcept;
+
+void __stdcall __std_atomic_unwait_indirect(const void* const _Storage, unsigned long long& _Wait_context) noexcept;
+
+std::size_t __stdcall __std_atomic_get_spin_count(const bool _Is_direct) noexcept;
+
+enum _Atomic_spin_phase : unsigned long long {

```

Comment:
I'm not an expert here - is it problematic that this is always 64-bit even on 32-bit architectures? Except for the negated value below, the constants are 32-bit.

Reply:
Yes, that negated value is why I want 64-bit. 
Slightly less efficient code, but makes counter aliasing problem from hypothetical to more hypothetical.

This is for counter of generation, it is incremented in 0x10 increments, and at some point wraps. If counter is 32-bit, it wraps on 0xFFFF'FFF0, if it is 64-bit it wraps on 0xFFFF'FFFF'FFFF'FFF0. Wrap may happen if waiting thread is preempted in very unpleasant moment, and gains control back a lot of time later, enough to increment counter exactly that many times to make it the same as before. This causes loss wakeup.

@ogiroux made in his implementation waiting for few seconds instead of forever to compensate against this situation. **I want to avoid such solution at all costs**, since it is highly desirable that waiting code can safely page out, and not page int.

Such non-strict solutions, where counter hypothetically may overlap, are common for synchronization algorithms, for example for SeqLock, or to prevent ABA problem. 

I don't see easy way to avoid it, so I want a counter to be large type.

Reply:
Now as I've implemented timed wait and changed context to be a structure, so now this counter is not shared with indirect counter, so it is now `unsigned long`